sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -183,6 +188,45 @@ getOrBuild(KernelProgramCache &KPCache, GetCachedBuildFT &&GetCachedBuild,
 
     return BuildResult;
   } catch (const exception &Ex) {
+    // We ran out of resources, so we want to clear the cache.
+    // We need acquire exclusive access of all the maps and any
+    // objects they reference, and then we can clear the cache.
+    if (Ex.get_cl_code() == PI_ERROR_OUT_OF_RESOURCES) {
+      if (n_retry == 1) {
+        std::rethrow_exception(std::current_exception());
+      }
+      // We need all the threads in the first loop of this function
+      // to break out that loop, so we'll notify them with by setting
+      // BS_WillEvict. Also, we don't want them to go back into the loop
+      // until we clear the cache, so we'll lock the maps, then we'll
+      // notify the waited build results.
+      std::lock_guard L1(KPCache.MProgramCacheMutex);
+      std::lock_guard L2(KPCache.MKernelsPerProgramCacheMutex);
+      std::lock_guard L3(KPCache.MKernelFastCacheMutex);
+      {
+        std::lock_guard<std::mutex> Lock(BuildResult->MBuildResultMutex);
+        BuildResult->State.store(BuildState::BS_WillEvict);
+      }
+
+      KPCache.notifyAllBuild(*BuildResult);
+      // TODO: possible but unlikely race condition - in the first
+      // loop, the threads will wait until the state of the BR
+      // changes, but will reread the state to make a decision on what
+      // to do next. I believe it is theoretically allowable for this
+      // clearCache to not be inter-thread happen the reread.
+
+      // TODO: possible race condition - the value returned by this
+      // function is a pointer to a BR. Clearing the cache will
+      // invalidate this pointer.  In the PiProgram case, we can just
+      // the return this handle stored in the BR directly. In the
+      // PiKernel case, it's a little tricker, as other parts of the
+      // runtime (i.e. enqueueImpKernel) use a pointer to the mutex
+      // that is stored in the BR. A shared pointer should circumvent
+      // this.
```

Comment:
In a similar vein, what happens to builds that might still be ongoing but have not failed due to limited resources? Will they be evicted with the rest? If so, maybe `clearCache` should have a mode that let these stay in the cache?